### PR TITLE
[GEOT-7924] Cascaded WMS regression on unencoded UTF-8 characters in the capabilities URLs

### PR DESCRIPTION
**Pull request by aaime**
[GEOT-7924](https://osgeo-org.atlassian.net/browse/GEOT-7924) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)  

See ticket for details

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [x] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java [hunk 1]
```diff
@@ -131,6 +131,114 @@ protected static URL appendURL(URL oldUrl, Map<String, String> appendQuery) thro
...
     }
 
+    /**
+     * Percent-encodes the illegal characters (non ASCII and disallowed ASCII) in the path, query and fragment of the
+     * given URL, leaving legal characters and existing {@code %XX} escapes untouched (a stray {@code %} becomes
+     * {@code %25}). This restores the automatic request line encoding Apache HttpClient 4 did and that HttpClient 5 and
+     * {@link URLConnection} do not: without it a capabilities document exposing an online resource with raw UTF
+     * characters breaks cascading. The scheme and authority are kept verbatim: a non-ASCII host is not percent-encoded
+     * but IDNA-encoded to ASCII punycode (see RFC 5890), encoding it here would produce an unresolvable name.
+     */
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java:142]: On the host: the javadoc says the opposite of what the comment reads into it. The host is left exactly as it is, and the javadoc explains why — percent escapes are the 
wrong tool for a host name, a non-ASCII host needs punycode instead.

File: modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java [hunk 2]
```diff
@@ -131,6 +131,114 @@ protected static URL appendURL(URL oldUrl, Map<String, String> appendQuery) thro
...
+    protected static URL encodeURL(URL url) throws MalformedURLException {
+        String file = url.getFile();
+        String ref = url.getRef();
+        String encFile = encodeIllegalCharacters(file);
+        String encRef = ref == null ? null : encodeIllegalCharacters(ref);
+        if (encFile.equals(file) && (ref == null || encRef.equals(ref))) {
+            return url;
+        }
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java:152]: On the fast path: the code already does what is suggested. encodeIllegalCharacters creates the StringBuilder only when it meets the first 
character that needs encoding, and returns the original string when it meets none; encodeURL then hands back the same URL object if nothing changed. All that is left is reading path and query once, and there
is no way around that — to know that no character needs encoding you have to look at all of them.

File: modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java [hunk 3]
```diff
@@ -131,6 +131,114 @@ protected static URL appendURL(URL oldUrl, Map<String, String> appendQuery) thro
...
+                if (i + 2 < segment.length() && isHex(segment.charAt(i + 1)) && isHex(segment.charAt(i + 2))) {
+                    sb.append('%');
+                } else {
+                    appendUtf8PercentEncoded(sb, '%');
+                }
+            } else if (cp < 128 && isLegalUriChar((char) cp)) {
+                sb.append((char) cp);
+            } else {
+                appendUtf8PercentEncoded(sb, cp);
+            }
```
[the remark this answers is not in the document]
Comment (aaime) [reply modules/library/http/src/main/java/org/geotools/http/AbstractHttpClient.java:193]: The edge cases have tests, in AbstractHttpClientTest: testStrayPercent checks that 50%, %zz and a trailing % all turn into %25, and testLoneSurrogateBecomesReplacementChar checks that a broken half character
turns into %EF%BF%BD instead of bytes no server could read. That file does not appear in the list of reviewed files, so it seems it was never opened — the note at the top of the review says the full review 
could not run because no runner was free.
